src/main/java/org/wso2/carbon/apimgt/securityenforcer/utils/SecurityUtils.java

```diff
@@ -0,0 +1,299 @@
+                            AISecurityException.CLIENT_REQUEST_ERROR_MESSAGE);
+                }
+            }
+
+            for (String headerKey : headerKeysSet) {
+                String headerValue = transportHeadersMap.get(headerKey);
+                transportHeadersArray.add(addObj(headerKey, headerValue));
+            }
+            return transportHeadersArray;
+        } else {
+            log.error("No Transport headers found for the request " + correlationID);
+            throw new AISecurityException(AISecurityException.CLIENT_REQUEST_ERROR,
+                    AISecurityException.CLIENT_REQUEST_ERROR_MESSAGE);
+        }
+    }
+
+    public static JSONObject addObj(String key, Object value) {
+        JSONObject obj = new JSONObject();
+        obj.put(key, value);
+        return obj;
+    }
+
+    /**
+     * Return a CloseableHttpClient instance
+     *
+     * @param protocol- service endpoint protocol. It can be http/https
+     * @param dataPublisherConfiguration - DataPublisher Configurations
+     *          maxPerRoute- maximum number of HTTP connections allowed across all routes.
+     *          maxOpenConnections- maximum number of HTTP connections allowed for a route.
+     *          connectionTimeout- the time to establish the connection with the remote host
+     *
+     * @return CloseableHttpClient
+     */
+    public static CloseableHttpClient getHttpClient(String protocol,
+            AISecurityHandlerConfig.DataPublisherConfig dataPublisherConfiguration) throws AISecurityException {
+
+        PoolingHttpClientConnectionManager pool;
+        try {
+            pool = SecurityUtils.getPoolingHttpClientConnectionManager(protocol);
+        } catch (Exception e) {
+            throw new AISecurityException(e);
+        }
+
+        pool.setMaxTotal(dataPublisherConfiguration.getMaxOpenConnections());
+        pool.setDefaultMaxPerRoute(dataPublisherConfiguration.getMaxPerRoute());
+
+        //Socket timeout is set to 10 seconds addition to connection timeout.
+        RequestConfig params = RequestConfig.custom()
+                .setConnectTimeout(dataPublisherConfiguration.getConnectionTimeout() * 1000)
+                .setSocketTimeout((dataPublisherConfiguration.getConnectionTimeout() + 10) * 10000).build();
+
+        return HttpClients.custom().setConnectionManager(pool).setDefaultRequestConfig(params).build();
+    }
+
+    /**
+     * Return a PoolingHttpClientConnectionManager instance
+     *
+     * @param protocol- service endpoint protocol. It can be http/https
+     * @return PoolManager
+     */
+    private static PoolingHttpClientConnectionManager getPoolingHttpClientConnectionManager(String protocol)
+            throws KeyStoreException, NoSuchAlgorithmException, KeyManagementException, IOException,
+            CertificateException {
+
+        PoolingHttpClientConnectionManager poolManager;
+        if ("https".equals(protocol)) {
```

Comment:
better to move these to constants

Reply:
https://github.com/wso2-extensions/apim-handler-pingai/pull/22